src/telegram/process_update_task.rs

```diff
@@ -25,15 +26,90 @@ impl ProcessUpdateTask {
         Self { update }
     }
 
-    pub async fn process(&self) {
-        // TODO: process users messages here
-        // Create a ApiClient
+    pub async fn process(&self) -> Result<(), BotError> {
+        log::info!("Received a message {:?}", self.update);
+
         let api = AsyncApi::new(&RUST_TELEGRAM_BOT_TOKEN);
-        let opwm_token = &OPEN_WEATHER_MAP_API_TOKEN;
+        // for now with _ to pass warnings
+        let _opwm_token = &OPEN_WEATHER_MAP_API_TOKEN;
         let keypair = Rsa::private_key_from_pem(&BINARY_FILE).unwrap();
         let keypair = PKey::from_rsa(keypair).unwrap();
 
-        log::info!("Received a message {:?}", self.update);
+        if let UpdateContent::Message(message) = &self.update.content {
+            let (chat_id, user_id, user) = Self::get_info_from_message(message);
+
+            Self::send_typing(message, &api).await?;
+
+            let state = Self::fetch_state(&chat_id, user_id, user.clone(), &keypair).await?;
+
+            match state {
+                ClientState::Initial => match message.text.as_deref() {
+                    Some("/start") | Some("/start@RustWeather77Bot") => {
+                        Self::start(message, &user, &api).await?;
+                    }
+                    _ => {}
+                },
+
+                ClientState::Pattern => match message.text.as_deref() {
+                    Some("/pattern") | Some("/pattern@RustWeather77Bot") => {}
+                    _ => {}
+                },
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn fetch_state(
+        chat_id: &i64,
+        user_id: u64,
+        user: String,
+        keypair: &PKey<Private>,
```

Comment:
why are you passing it here? since this key is only used here, you can use directly in this function

Reply:
Done